File: Kısım2C/src/ImeiNo.c
#include "ImeiNo.h"

ImeiNo ImeiNoOlustur()
{
    ImeiNo this=(ImeiNo)malloc(sizeof(struct IMEINO));
    this->diziUzunluk=15;
    this->imei=(int*)malloc(sizeof(int)*this->diziUzunluk);    //diziler heap bölgesinde oluşturuluyor
    this->carpimVeToplamlar=(int*)malloc(sizeof(int)*this->diziUzunluk);
    this->imeiKontrol=(int*)malloc(sizeof(int)*this->diziUzunluk);
    this->rastgele=RastgeleOlustur();
    this->imeiNoGetir=&_imeiNoGetir;
    this->imeiNoKontrol=&_imeiNoKontrol;
    this->ImeiNoYoket=&_ImeiNoYoket;
    this->toplam=0;
    this->gecerliImeiSayac=0;
    this->gecersizImeiSayac=0;
return this;
}
void _imeiUret(const ImeiNo this) //private
{
    for (int i = 0; i < 14; ++i) {
        this->imei[i]=this->rastgele->randomDegerGetir(this->rastgele)%10;
        this->carpimVeToplamlar[i]=this->imei[i];
    }
    for (int i = 1; i < 14; i+=2)
    {
        this->carpimVeToplamlar[i]=this->carpimVeToplamlar[i]*2;    //tek basakmaklar ikiyle çarpılıyor ve sayı ondan büyükse rakamlar toplanıyor
        int birlerBasamagi=this->carpimVeToplamlar[i]%10;
        int onlarBasamagi=this->carpimVeToplamlar[i]/10;
        this->carpimVeToplamlar[i]=birlerBasamagi+onlarBasamagi;
    }
    for (int i = 0; i < 14; i++)
    {
        this->toplam+=this->carpimVeToplamlar[i];     //çarpılan sayılar toplanıyor
    }
    this-> toplam=( this->toplam*9)%10;  //15. basamak belirleniyor
    this->imei[14]= this->toplam;
    this->toplam=0;
}
void imeiKuralKontrolEt(const ImeiNo this) //private
{
   int toplam=0;
    for (int i = 1; i < 14; i+=2)
    {
        this->imeiKontrol[i]=this->imeiKontrol[i]*2;
        int birlerBasamagi=this->imeiKontrol[i]%10;
        int onlarBasamagi=this->imeiKontrol[i]/10;
        this->imeiKontrol[i]=birlerBasamagi+onlarBasamagi;
    }
    for (int i = 0; i < 15; i++)
    {
        toplam+=this->imeiKontrol[i];
    }
    if(toplam%10==0)
        this->gecerliImeiSayac++;

    else
        this->gecersizImeiSayac++;
}
void dosyaOkuBasamakSayisiBul(const ImeiNo this) //private
{                                    //bu fonksiyonda dosya okunuyor ve imei numarasının basamak sayısı kontrol ediliyor. Ardından check digit kontrolü için diger fonksiyon cagriliyor
    int bayrak=0;
    int index=0;
    FILE * fp = fopen("kisiler.txt", "r");
    char c;
    int kelimeSayac = 0;
    while((c = fgetc(fp)) != EOF)
    {
        if(c == ' ' || c == '\n' )
        {
            if((kelimeSayac-5)%6==0 &&bayrak==1)
            {
                imeiKuralKontrolEt(this);
            }
            else if((kelimeSayac-5)%6==0 &&bayrak==0)
                this->gecersizImeiSayac++;
            kelimeSayac++;
            index=0;
        }
        else if(c!='(' && c!=')')
        {
            if((kelimeSayac-5)%6==0)
            {
                if(index<=14)
                {
                    this->imeiKontrol[index]=c-'0';
                    index++;
                    bayrak=1;
                }
                else
                {
                    bayrak=0;
                }
            }
        }
    }
    fclose(fp);
}
int* _imeiNoGetir(const ImeiNo this) //public
{
    _imeiUret(this);
    return this->imei;
}
void _imeiNoKontrol(const ImeiNo this) //public fonksiyon
{
	this->gecerliImeiSayac=0;
	this->gecersizImeiSayac=0;
    dosyaOkuBasamakSayisiBul(this);
    printf("Gecerli Imei Numarasi Sayisi: %d\n",this->gecerliImeiSayac);
    printf("Gecersiz Imei Numarasi Sayisi: %d\n",this->gecersizImeiSayac);
}
void _ImeiNoYoket(ImeiNo this)
{
    if(this==NULL)return;
    this->rastgele->RastgeleYoket(this->rastgele);
    free(this->imei);
    free(this->carpimVeToplamlar);
    free(this->imeiKontrol);
    free(this);

}

File: Kısım1Java/Kutuphane/RASTGELEKISIURET/src/MyPackage/IMEINo.java
/**
*
* @author hakandurgay@gmail.com hakan durgay
* @since 26.04.2020
* <p>
* Açıklama
* </p>
*/
package MyPackage;

import java.io.FileNotFoundException;
import java.util.Scanner;

/**
 *
 * @author Hakan Durgay
 */
public class  IMEINo{
  private  int [] imei;
  private  int [] carpimVeToplamlar;
  private int [] imeiKontrol;
  private int toplamKontrol;
  private int toplam;
  private int gecerliImeiSayisi;
  private int gecersizImeiSayisi;
  private String kontrolimei;
  private String imeiNo;
  Rastgele rand;
 IMEINo()
 {
        toplam=0;
        gecerliImeiSayisi=0;
        gecersizImeiSayisi=0;
        rand=new Rastgele(10);
        imei=new int[15];
        carpimVeToplamlar=new int[15];
        imeiKontrol=new int[16];
 }
           
    public String imeiNoGetir()
    {
        imeiNo="";
        imeiUret();
        return imeiNo;
    }
    public void imeiNoKontrol() throws FileNotFoundException
    {
        gecersizImeiSayisi=0;
        gecerliImeiSayisi=0;
        imeiKontrolEt();
        System.err.println("Gecerli Imei No Sayisi: "+gecerliImeiSayisi);
                System.err.println("Gecersiz Imei No Sayisi: "+gecersizImeiSayisi);

    }
    private void imeiUret()
    {
        for (int i = 0; i < 14; i++)
        {
            imei[i]=(int)rand.randomDegerGetir();
            carpimVeToplamlar[i]=imei[i];     //ilk hali bozulmaması için başka diziye atılıyor
        }
        for (int i = 1; i < 14; i+=2) 
        {
            carpimVeToplamlar[i]=carpimVeToplamlar[i]*2;    //tek basakmaklar ikiyle çarpılıyor ve sayı ondan büyükse rakamlar toplanıyor
            int birlerBasamagi=carpimVeToplamlar[i]%10;
            int onlarBasamagi=carpimVeToplamlar[i]/10;
            carpimVeToplamlar[i]=birlerBasamagi+onlarBasamagi;
        }
        for (int i = 0; i < 14; i++) 
        {
             toplam+=carpimVeToplamlar[i];     //çarpılan sayılar toplanıyor
        }
        toplam=(toplam*9)%10;  //15. basamak belirleniyor
        imei[14]=toplam;
                toplam=0;
        for (int i = 0; i < 15; i++)
        {   
          imeiNo+=imei[i];  //oluşan imei stringe atılıyor
        }    
    }
    private void imeiKontrolEt() throws FileNotFoundException
    {
        char ch;
        int sayac=0;
        java.io.File dosyaOku = new java.io.File("Kisiler.txt");  //int deger okuma 
          Scanner oku = new Scanner(dosyaOku);

          while (oku.hasNext())
          {      
              if((sayac-5)%6==0) //imeinin bulunduğu sütunlar
              {
                   kontrolimei =oku.next(); //okunan değer stringe atılıyor
                   if(basamakSayisiKontrolEt(kontrolimei))       //imei numarasının basamak sayısı 15 ise check digit kontrolü için diziye atılıyor
                   {
                       for (int i = 1; i < kontrolimei.length(); i++)
                         { 
                                  ch=kontrolimei.charAt(i);  //parantezlerden kurtulmak için döngü birden başlıyor
                                  int rakam = ch - '0'; //chardan inte dönüştürülüyor
                                     imeiKontrol[i-1]=rakam;                                                          
                         }
                       imeiNoDogrulama();
                   }
                   else
                   {
                       gecersizImeiSayisi++;
                   }
              }      
              else
              oku.next();
              sayac++;                        
          }     
          oku.close();
    }
    private boolean basamakSayisiKontrolEt(String imei)
    {
        int basamakSayaci=0;
        for (int i = 0; i < imei.length()-2; i++) 
        {
            basamakSayaci++;
        }
        if(basamakSayaci!=15)
            return false;
        else
            return true;
    }
    private void imeiNoDogrulama()
    {
        toplamKontrol=0;
        toplam=0;
        for (int i = 1; i < 14; i+=2) 
        {
            imeiKontrol[i]=imeiKontrol[i]*2;
            int birlerBasamagi=imeiKontrol[i]%10;
            int onlarBasamagi=imeiKontrol[i]/10;
            imeiKontrol[i]=birlerBasamagi+onlarBasamagi;
        }
        for (int i = 0; i < 15; i++) 
        {
             toplamKontrol+=imeiKontrol[i];  
        }
        if(toplamKontrol%10==0)    
         gecerliImeiSayisi++;
        else
        gecersizImeiSayisi++;

    } 
}
